Add plugin support for other templates languages

sarahelsaig opened the pull request:
I've thought this through, and I think Roslyn C# script plugins make the most sense. (also to make it work with DLLs you'd have to publish OrchardCoreContrib.PoExtractor.Abstractions to NuGet as a separate library package - probably not worth the effort) I've included a unit test and sample plugin that hopefully gets the point across without being too niche. (The important change is in the `Program.ProcessPluginsAsync` method, the rest is just plumbing.)

Fixes #101

hishamco commented:
I just looked into the PR and noticed a dependency on `Program.cs` which is wrong for me. IMHO exposing the `IProjectProcessor` will make the tool extendable. I am still thinking if we need the `-p` or if we can make the tool open some possibilities for the others to add their own options

Let me hear your feedback on this

sarahelsaig commented:
> I just looked into the PR and noticed a dependency on Program.cs which is wrong for me.

If you mean the `ScriptOptions.Default.AddReferences(typeof(Program).Assembly)` [here](https://github.com/OrchardCoreContrib/OrchardCoreContrib.PoExtractor/pull/102/files#diff-278f797f65a94d95062662ff3b79e00916d305d4bf7e41cfbb13224c9d4d2373R139), I've only used `Program` to expose all the libraries to the script the tool has already loaded. As I see it, this is not a security application so we can be very permissive.
For example this way the script has access to `OrchardCoreContrib.PoExtractor.DotNet.CS` too, not just `OrchardCoreContrib.PoExtractor.Abstractions`. If someone wants to have a different take on one of the existing processors, this could be a way to have your own custom (more opinionated) solution without it affecting the main repo, while re-using the auxiliary classes that are already in those projects. I don't see the harm in it, so could you explain why it's wrong for you?

> IMHO exposing the IProjectProcessor will make the tool extendable.

I think it's not possible to expose just one type, you have to expose its whole assembly. We can change the above linked code to only expose `OrchardCoreContrib.PoExtractor.Abstractions` instead of everything, but that feels unnecessarily restrictive.

> I am still thinking if we need the -p or if we can make the tool open some possibilities for the others to add their own options.

Sorry, I don't really understand. This `-p` approach is the simplest way to add broad extensibility that I could think of, while being less maintenance in your code base. If you want something more elaborate without the need to use a command line switch we can discuss, but I'm not sure if it's really necessary.

sarahelsaig commented:
Also by importing `Program`'s assembly the plugin developer will have access to all the of external dependencies PoExtractor has already loaded. For example if I only included the `OrchardCoreContrib.PoExtractor.Abstractions`, then the example plugin will fail with the following error:
> error CS0234: The type or namespace name 'Json' does not exist in the namespace 'System.Text' (are you missing an assembly reference?)

So now I'd have to include `#r "nuget: System.Text.Json, 9.0.0"`  at the start of the script, and NuGet may have to download this version if it gets out of sync from what the tool itself uses. This is annoying imo.

sarahelsaig commented:
@hishamco Have you seen my replies? This is not super urgent yet but it would be good if we could have a solution before year's end, because we might need it next month.

hishamco commented:
I didn't look into the last comment, but the PR is opened in one tab, so I need to revise it today or tomorrow to proceed

sarahelsaig commented:
Hi @hishamco , any news with this?

hishamco commented on NuGet.config:
```diff
@@ -1,8 +1,8 @@
     <clear />
-    <add key="NuGet" value="https://api.nuget.org/v3/index.json" />
-    <add key="OrchardCore" value="https://nuget.cloudsmith.io/orchardcore/preview/v3/index.json" />
+    <add key="nuget.org" value="https://api.nuget.org/v3/index.json" />
```
The feed is already there?!!

sarahelsaig replied:
Ooops, I didn't notice. I've reverted the name.

My point was just to remove the `OrchardCore` source. It's not even used and because of the central package management this kept giving `warning NU1507: There are 2 package sources defined in your configuration.`

hishamco commented on src/OrchardCoreContrib.PoExtractor.Abstractions/PluginHelper.cs:
```diff
@@ -0,0 +1,25 @@
+        foreach (var plugin in plugins)
+        {
+            var code = await File.ReadAllTextAsync(plugin);
+            await CSharpScript.EvaluateAsync(code, options, new PluginContext(projectProcessors, projectFiles));
```
I'm not sure why this in Abstractions while it relies heavily on `CSharpScript`

sarahelsaig replied:
Where should it be then?

sarahelsaig replied:
The reason is that this class only relies on `IProjectProcessor` from this project, which is in `OrchardCoreContrib.PoExtractor.Abstractions`, so I [imported `CSharpScript`'s package](https://github.com/OrchardCoreContrib/OrchardCoreContrib.PoExtractor/pull/102/files#diff-ced54314a7ece4354e01069187749b6edec9c099b9d27f1ccb43e1d2cef73d66R6) there as well. I think it's appropriate, because this helper is related to extensibility. We can move it into `OrchardCoreContrib.PoExtractor` if you prefer.

sarahelsaig replied:
I've moved it to `OrchardCoreContrib.PoExtractor`.

sebastienros replied:
I am not familiar with POe but anything that is related to c# plugins should be in a separate project or in the main assembly, not abstraction. No reference, no class.

hishamco commented:
@sarahelsaig Sorry for the late, I revised the code one more time still worried about the impact of dynamic linking, so I'm thinking for two simple approaches:

1. Creating `OrchardCoreContrib.PoExtractor.VueJs` directly in the repo
2. Reference `Lombiq.VueJs` to `OrchardCoreContrib.PoExtractor`

While this repo is Orchard Core Contrib it SHOULD help any OC project, I might think the first option is better, but I need to hear your feedback. Also, I will think for the last time about dynamic linking and its drawbacks

sarahelsaig commented:
Can you explain what is your worry about dynamic linking? What are the drawabacks? I don't see any impact unless you manually opt-in by using the command line switch. Otherwise I think it's safe with only very minimal overhead. I see it as a transparent improvement.

> this repo is Orchard Core Contrib it SHOULD help any OC project

I fully agree with this sentiment. Note that since my original PR, I have seen other highly project-specific opportunities where this feature could be useful. In both of the cases below I could use a PoExtractor plugin to centralize other localization tech so OC can manage the actual localization and then serve it back to the associated tech. This will reduce the number of tools for the staff to learn to do the translations, and make complete translation management easier and safer.
- We have a client project that uses OC as a headless server with a complementing node.js frontend. We could harvest the english l18n JSON files used by the frontend (similarly to the example plugin in this PR) and then serve back the other languages.
- In the same project we also have to use a XSLT to generate some visualizations (it's a legal requirement). For now we manually maintain XML translation files for the languages we need in the MVP, but having a third localization tech in the project just for one small feature is very annoying.

I don't think we should create OrchardCoreContrib.PoExtractor.VueJs, not other OrchardCoreContrib.PoExtractor projects for the use-cases I mentioned above. This would quickly bloat your repo.

> Reference Lombiq.VueJs to OrchardCoreContrib.PoExtractor

I'm not sure if I understand. If you mean using OrchardCoreContrib.PoExtractor as a library inside Lombiq.VueJs, your existing NuGet tool package doesn't make it possible. Even if you released a NuGet library package it would be pointlessly cumbersome to use in my opinion.

sarahelsaig commented:
@hishamco Happy New Year! Have seen my last comment?

Even if you don't want to go forward with this PR I'd like to understand what's your concern about using C# scripts in this project. Maybe you are right and I should avoid using it in my own projects too.

Also just to clear things up, since you mentioned "dynamic linking" in your previous comment. The topic came up in the issue, but in this PR I did not implement dynamic linking. What we have here is just scripting. Similar to using Jint in OrchardCore, except the script language is also C#.

hishamco commented:
Happy New Year!

Let us finalize the PR this week, it would be nice to demo it in the upcoming meeting if it's possible I need to have @sebastienros feedback

sarahelsaig commented:
@hishamco I've demoed it yesterday, it was received positively. Sébastien gave some suggestions: to add the ability to use a remote script via URL, and to ensure that you can load in external libraries from the script. I have addressed these and your last review comments in my latest commits.

I hope we can resolve this PR soon. :slightly_smiling_face:

hishamco commented:
Sound good, waiting for the recording if you finalize reacting to Seb's comment let us merge it this week

hishamco commented on README.md:
```diff
@@ -42,6 +42,24 @@ Specifies the code language to extracts translatable strings from. Default: `C#`
 
+- **`-p|--plugin {path or URL to CSX file}`**
+
+Specifies a path to a C# script file which can define further project processors. (You can find an example script [here](test/OrchardCoreContrib.PoExtractor.Tests/PluginTestFiles/BasicJsonLocalizationProcessor.csx).) This can be used to process localization from code languages or template engines not supported by the above options. You can have multiple of this switch in one call to load several  ...
```
```suggestion
Specifies a path to a C# script file that can define further project processors. (You can find an example script [here](test/OrchardCoreContrib.PoExtractor.Tests/PluginTestFiles/BasicJsonLocalizationProcessor.csx).) This can be used to process localization from code languages or template engines not supported by the above options. You can have multiple of these switches in one call to load several plugins simultaneously. If the argument starts with `https://` then it's treated as a web URL and the script at that address is downloaded into memory and executed instead of a local file.
```

hishamco commented on NuGet.config:
```diff
@@ -1,8 +1,8 @@
+    <!-- Ignore global configuration -->
     <clear />
     <add key="NuGet" value="https://api.nuget.org/v3/index.json" />
-    <add key="OrchardCore" value="https://nuget.cloudsmith.io/orchardcore/preview/v3/index.json" />
```
Why?

hishamco commented:
I'm planning to use CLI parsing package to make things easier, let me create a quick PR to use https://github.com/natemcmaster/CommandLineUtils the you can react to the changes, then we could merge

Hope the recording for yesterday's meeting will be published ASAP, I need to have a look before we merge this. No worry we will try to merge this PR this week

sarahelsaig commented:
I have rewritten `Program` to work with your new command line argument system. I have also found and fixed [a bug](https://github.com/OrchardCoreContrib/OrchardCoreContrib.PoExtractor/pull/102/commits/284e8b35b165887a72fb7568462f6791ae139bfa) that we missed in the previous PR.

About the recording, maybe it will be faster if you ask for a direct link from someone who is part of the main Orchard org.

hishamco commented:
> I have rewritten Program to work with your new command line argument system. I have also found and fixed [a bug](https://github.com/OrchardCoreContrib/OrchardCoreContrib.PoExtractor/pull/102/commits/284e8b35b165887a72fb7568462f6791ae139bfa) that we missed in the previous PR.

I'm suggesting a new PR for it or I will do it so quick, we need to focus on what it's related to the plugin support

Thanks again

sarahelsaig commented:
> I'm suggesting a new PR for it or I will do it so quick, we need to focus on what it's related to the plugin support
> 
> Thanks again

Sure, here is the PR: https://github.com/OrchardCoreContrib/OrchardCoreContrib.PoExtractor/pull/104

hishamco commented:
I'm still waiting for Gabor to check the recording, but I will do one more review tonight

hishamco commented:
I just saw the recording and heard @sebastienros feedback:

- Shall we need to support absolute & relative URLs for plugins
- @sebastienros from a security perspective do you think this is safe, especially if someone wrote a C# plugin

Last not Liquid is already supported and the tool is able to extract the strings  https://github.com/OrchardCoreContrib/OrchardCoreContrib.PoExtractor/tree/main/src/OrchardCoreContrib.PoExtractor.Liquid

hishamco commented on src/OrchardCoreContrib.PoExtractor/PluginHelper.cs:
```diff
@@ -0,0 +1,55 @@
+        assemblies ??=
+        [
+            typeof(IProjectProcessor).Assembly, // OrchardCoreContrib.PoExtractor.Abstractions
+            typeof(ExtractingCodeWalker).Assembly, // OrchardCoreContrib.PoExtractor.DotNet
```
Why this is needed?!!

sarahelsaig replied:
I wanted to add access to the full assembly, in case the plugin writer wants to derive a type from something existing. Like the way `RazorProjectProcessor` inherits from `CSharpProjectProcessor`.

But I just checked and [only adding `Program`'s assembly](https://github.com/OrchardCoreContrib/OrchardCoreContrib.PoExtractor/pull/102/commits/04f66300831ec28999bedfb3be4f529a6e9b1631) is enough. So I've updated the code accordingly.

hishamco commented on src/OrchardCoreContrib.PoExtractor/PluginHelper.cs:
```diff
@@ -0,0 +1,55 @@
+        }
+    }
+
+    public record PluginContext(List<IProjectProcessor> projectProcessors, List<string> projectFiles);
```
```suggestion
    public record PluginContext(List<IProjectProcessor> ProjectProcessors, List<string> ProjectFiles);
```

sarahelsaig replied:
In the CSX code the properties of this object become global variables. In other samples I have seen, such variables used camelCase, so I figured it's the convention.

Please confirm if you still want it turned into PascalCase with this information in mind. Or add a comment explaining why they are camelCase.

hishamco replied:
I got VS naming violation

sarahelsaig replied:
Strange, I didn't get any. Okay, changing it to PascalCase then.

hishamco commented on src/OrchardCoreContrib.PoExtractor/Program.cs:
```diff
@@ -102,7 +121,7 @@ public static void Main(string[] args)
                 }
 
-                if (isSingleFileOutput)
+                if (!isSingleFileOutput)
```
This is fixed in another PR, right?

sarahelsaig replied:
Yes. I merged from your `main` branch and now it's not shown as a diff in this PR.

hishamco commented on test/OrchardCoreContrib.PoExtractor.Tests/PluginTests.cs:
```diff
@@ -0,0 +1,72 @@
+        Assert.Equal(CleanupSpaces(expectedResult), CleanupSpaces(actualResult));
+    }
+
+    private static string CleanupSpaces(string input)
+    {
+        // Trim leading whitespaces.
+        input = Regex.Replace(input.Trim(), @"^\s+", string.Empty, RegexOptions.Multiline);
+        
+        // Make the path OS-specific, so the test works on Windows as well. 
+        return input.Replace('/', Path.DirectorySeparatorChar);
+    }
```
```suggestion
    private static string CleanupSpaces(string input)
    {
        // Trim leading whitespaces.
        input = TrimLeadingSpacesRegex().Replace(input.Trim(), string.Empty);
        
        // Make the path OS-specific, so the test works on Windows as well. 
        return input.Replace('/', Path.DirectorySeparatorChar);
    }
    
    [GeneratedRegex(@"^\s+", RegexOptions.Multiline)]
    private static partial Regex TrimLeadingSpacesRegex();
```

hishamco commented:
I did a last review, please check and let merge then

hishamco commented:
Please comment on https://github.com/OrchardCoreContrib/OrchardCoreContrib.PoExtractor/pull/102#discussion_r1924480041 then we can merge so quick

hishamco commented:
Thanks so much @sarahelsaig, I think the next step is to publish the abstractions project

hishamco commented:
BTW did you start a VueJS localization extractor, or it's something private for Lombiq

sarahelsaig commented:
I will start working on it soon (https://github.com/Lombiq/Orchard-Vue.js/issues/152).
It will be open source in the https://github.com/Lombiq/Orchard-Vue.js/ repo.
